protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol

```diff
@@ -193,111 +311,96 @@ contract Weather is Sun {
-     * as the twa reserves are not indiciative of the current deltaB in the pool.
-     *
-     * Generalized for a single well. Sop does not support multiple wells.
+    /*
+     * @notice Calculates the amount of beans per well that should be minted in a sop.
+     * @param wellDeltaBs The deltaBs of all whitelisted wells in which to flood. Must be sorted in descending order.
      */
-    function calculateSop(address well) private view returns (uint256 sopBeans, IERC20 sopToken) {
-        // if the sopWell was not initalized, the should not occur.
-        if (well == address(0)) return (0, IERC20(address(0)));
-        IWell sopWell = IWell(well);
-        IERC20[] memory tokens = sopWell.tokens();
-        Call[] memory pumps = sopWell.pumps();
-        IInstantaneousPump pump = IInstantaneousPump(pumps[0].target);
-        uint256[] memory instantaneousReserves = pump.readInstantaneousReserves(
-            well,
-            pumps[0].data
-        );
-        uint256[] memory currentReserves = sopWell.getReserves();
-        Call memory wellFunction = sopWell.wellFunction();
-        (uint256[] memory ratios, uint256 beanIndex, bool success) = LibWell.getRatiosAndBeanIndex(
-            tokens
-        );
-        // If the USD Oracle oracle call fails, the sop should not occur.
-        // return 0 rather than revert to prevent sunrise from failing.
-        if (!success) return (0, IERC20(address(0)));
-
-        // compare the beans at peg using the instantaneous reserves,
-        // and the current reserves.
-        uint256 instantaneousBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(instantaneousReserves, beanIndex, ratios, wellFunction.data);
-
-        uint256 currentBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(currentReserves, beanIndex, ratios, wellFunction.data);
+    // code review note: have some casual comments here for just understanding the code, happy to
+    // change this function to make it more efficient based on feedback.
+    function calculateSopPerWell(
+        WellDeltaB[] memory wellDeltaBs,
+        uint256 totalPositiveDeltaB,
+        uint256 totalNegativeDeltaB,
+        uint256 positiveDeltaBCount
+    ) public pure returns (WellDeltaB[] memory) {
+        // most likely case is that all deltaBs are positive
+        if (positiveDeltaBCount == wellDeltaBs.length) {
+            // if all deltaBs are positive, need to sop all to zero, so return existing deltaBs
+            return wellDeltaBs;
+        }
 
-        // Calculate the signed Sop beans for the two reserves.
-        int256 lowestSopBeans = int256(instantaneousBeansAtPeg).sub(
-            int256(instantaneousReserves[beanIndex])
-        );
-        int256 currentSopBeans = int256(currentBeansAtPeg).sub(int256(currentReserves[beanIndex]));
+        if (positiveDeltaBCount == 0) {
```

Comment:
can we combine the two if statements given that the logic executed is exactly the same? I'd also the `totalPositiveDeltaB < totalNegativeDeltaB` conditional checked first, given that its marginally more likely to occur (which will cause the if statement to not check the 2nd conditional)

```suggestion
        if (positiveDeltaBCount == 0 || totalPositiveDeltaB < totalNegativeDeltaB) {
```

Reply:
Combined here: https://github.com/BeanstalkFarms/Beanstalk/pull/870/commits/8bfa314d12ccf77bf815e986fcfc2bf0b6572f22

Reply:
lgtm